### PR TITLE
[Feature] Adds DynamoDB connection

### PR DESCRIPTION
**Pull request by Sqyphen**
## Description
Adds connection details for DynamoDB

## Implementation
Adds new paths to router to create new debates and to get existing debates

## Screenshot
/n/a

## .env changes
There are new .env changes, please make sure you are up to date with the order order .env details in Last pass

## Additional requirements
Run `npm install` on pulling this branch to get the latest packages

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -309,6 +309,22 @@
...
+        "events": "1.1.1",
+        "ieee754": "1.1.8",
+        "jmespath": "0.15.0",
+        "querystring": "0.2.0",
+        "sax": "1.2.1",
+        "url": "0.10.3",
+        "uuid": "3.3.2",
+        "xml2js": "0.4.19"
+      }
+  {,
```
Comment (Lily2point0) [inline package-lock.json]: Rogue comma, package lock is not working.

Comment (Lily2point0) [reply package-lock.json]: Also missing a closing bracket? Advise deleting and re-generating package-lock.

File: index.js [hunk 2]
```diff
@@ -16,6 +16,14 @@ const bodyParser = require("body-parser");
...
 const debate_routes = require("./routes/debate");
 const api_routes = require("./routes/api");
+const AWS = require("aws-sdk");
+
+const dynamoDb = new AWS.DynamoDB.DocumentClient({
+  region: "eu-west-1"
+});
+
+// view engine setup
+var hbs = require("express-hbs");
```
Comment (Lily2point0) [inline index.js]: `hbs` is already declared. also `const`, not sure how it got mixed up.

File: package-lock.json (hunk 1 above)
Comment (Sqyphen) [reply package-lock.json]: Re-generated and fixed

File: index.js [hunk 3]
```diff
@@ -52,13 +66,72 @@ app.get("/", (req, res) => {
...
+  } catch (err) {
+    console.error(err);
+    res.status(404).send("Sorry can't find that!");
+  }
+});
+
+app.get("/:debateType/:debateName", (req, res) => {
+  try {
+    const { debateName, debateType } = req.params;
+    const data = require(`./dummyData/${debateType}/${debateName}.json`);
```
Comment (Lily2point0) [inline index.js]: That is still using mock data, not Dynamo.

File: index.js (hunk 2 above)
Comment (Sqyphen) [reply index.js]: Not sure either
Fixed

File: index.js (hunk 3 above)
Comment (Sqyphen) [reply index.js]: Correct :)

Comment (Sqyphen) [reply index.js]: That hasn't been done yet, this branch was just to establish a connection with initial get and put paths

File: index.js [hunk 4]
```diff
@@ -52,13 +66,72 @@ app.get("/", (req, res) => {
...
+    const result = await dynamoDb.put(params).promise();
+    console.log(result);
+    res.send(JSON.stringify(result));
+  } catch (err) {
+    console.error(err);
+    res.status(404).send("Sorry can't find that!");
+  }
+});
+
+app.get("/debate", async (req, res) => {
```
Comment (Lily2point0) [inline index.js]: Overriden by another route, never getting to that stage.
Can't test

Comment (Sqyphen) [reply index.js]: Override has been fixed now

File: helpers/listings.js [hunk 5]
```diff
@@ -40,7 +45,25 @@ function getDebateListings(folder, searchedType = "") {
   return directoryList;
 }
 
+async function getDynamoDebateListings() {
+  /*
```
Comment (Lily2point0) [inline helpers/listings.js]: Why is this commented out? Delete if not needed -- can get it back from history

Comment (Sqyphen) [reply helpers/listings.js]: Ah, thanks, removed

File: static/js/admin/main.js [hunk 6]
```diff
@@ -54,7 +54,7 @@ function submitForm(e) {
   } else {
     try {
       var promise = new Promise((resolve, reject) => {
-        resolve(submitData("/api/create_new_debate", data));
+        resolve(submitData("/api/debate/create_new", data));
```
Comment (Lily2point0) [inline static/js/admin/main.js]: is `_new` necessary? `create` would do as it'd be `update` otherwise?

File: routes/api.js [hunk 7]
```diff
@@ -1,24 +1,122 @@
...
+    const params = {
+      TableName: process.env.DEBATE_TABLE,
+      ProjectionExpression: "debateType"
+    };
+    const result = await dynamoDb.scan(params).promise();
+    res.send(JSON.stringify(result));
+  } catch (err) {
+    console.error(err);
+    res.status(404).send(err);
+    //res.status(404).send("Sorry can't find that!");
```
Comment (Lily2point0) [inline routes/api.js]: Would leave a "human friendly" response here, as the error is likely to happen often (if the requested debate doesn't exist). Can leave the `err` as log.

File: static/js/admin/main.js (hunk 6 above)
Comment (Sqyphen) [reply static/js/admin/main.js]: Fair point, ammended

File: routes/api.js (hunk 7 above)
Comment (Sqyphen) [reply routes/api.js]: Updated 👍

File: routes/api.js [hunk 8]
```diff
@@ -1,24 +1,121 @@
...
+    console.log(result);
+    res.send(JSON.stringify(result));
   } catch (err) {
+    console.error(err);
     res.status(404).send("Sorry can't find that!");
   }
 });
 
-// TODO: add any other required api calls
+router.get("/debate", async (req, res) => {
```
Comment (Lily2point0) [inline routes/api.js]: Doesn't need to be a route

Comment (Sqyphen) [reply routes/api.js]: Agreed, removed

File: routes/api.js [hunk 9]
```diff
@@ -1,24 +1,77 @@
...
+  try {
+    const { name, seriesId } = req.params;
+    const debate = await dynamo_db.getById(name, seriesId);
+    res.send(JSON.stringify(debate));
   } catch (err) {
-    res.status(404).send("Sorry can't find that!");
+    console.error(err);
+    res
+      .status(404)
+      .send("Sorry can't find that! Issue with GET /debate/:name/:seriesId");
```
Comment (Lily2point0) [inline routes/api.js]: `Issue with /debate/${name}/${seriesId}` would be more helpful

File: routes/api.js [hunk 10]
```diff
@@ -1,24 +1,77 @@
...
+      });
+      res.end();
+      return;
+    }
 
-    // TODO: add submission to data storage
+    const timestamp = new Date().getTime();
+    const params = {
+      Item: {
+        name: data.name,
```
Comment (Lily2point0) [inline routes/api.js]: The primary key for DynamoDB must be a unique value. I can foresee issues having the name as primary key.

File: routes/api.js (hunk 9 above)
Comment (Sqyphen) [reply routes/api.js]: Done 👍

File: routes/api.js (hunk 10 above)
Comment (Sqyphen) [reply routes/api.js]: The combination of the name and seriesId is the Key, is that unique enough?

File: modules/hats.js [hunk 11]
```diff
@@ -1,12 +1,16 @@
 const comments = require("../helpers/comments");
 
 function render(req, res, data) {
```
Comment (Lily2point0) [inline modules/hats.js]: Can we pick another name than `render`, that's already used by hbs and might get confusing.

File: routes/api.js (hunk 10 above)
Comment (Lily2point0) [reply routes/api.js]: In what way? Do you mean as a value?

File: modules/hats.js (hunk 11 above)
Comment (Sqyphen) [reply modules/hats.js]: Sure I can change it 
Does hbs use `render`? I though it was Express?

Comment (Sqyphen) [reply modules/hats.js]: Done

**Pull request merged**